### DISCUSSION TITLE
Error with `in this` and `ref this` as Parameters in Static Method Calls

### DISCUSSION THREAD

**Discussion by aa89227** [dotnet/csharplang#8059] [Q&A] [answered]
Hello everyone,

I'm working on a C# 12 project and encountering errors when using in this and ref this as parameters in static method calls. Below is a snippet of the code that's causing the issues:

```cs
public class ClassA
{
    public void MethodA()
    {
        ClassWithStaticMethod.StaticMethodWithIn(this);
        
        // Error occurs here
        ClassWithStaticMethod.StaticMethodWithIn(in this);
        
        // Error occurs here
        ClassWithStaticMethod.StaticMethodWithRefReadonly(ref this);
        
        ClassWithStaticMethod.StaticMethodWithRefReadonly(this);
    }
}

public class ClassWithStaticMethod
{
    public static void StaticMethodWithIn(in ClassA classA)
    {
        Console.WriteLine("Static Method with in");
    }

    public static void StaticMethodWithRefReadonly(ref readonly ClassA classA)
    {
        Console.WriteLine("Static Method with ref readonly");
    }
}
```

When attempting to pass `this` with `in` and `ref readonly` modifiers in `StaticMethodWithIn` and `StaticMethodWithRefReadonly` respectively, the compiler throws errors. I would appreciate any insights on the following:

1. Why does the compiler not accept `in this` and `ref this` in these contexts?
2. Why does the compiler not throw any errors when just this is passed without any modifiers in these method calls?

Thank you for any explanations or workarounds you might suggest!

**Comment by iam3yal** [comment]
@aa89227 The question shouldn't be why the language not doing what you might expect or why the compiler enforces it but why do you think it should work? not to mention that both of the examples that do work implicitly pass reference by `ref`.

**Reply by aa89227** [reply]
I'm curious why passing `this` and `in this` results in different outcomes, given that both methods essentially pass a reference, yet the explicit keyword is not permitted.

**Comment by huoyaoyuan** [comment, marked as answer]
This is intentional and the whole purpose of `ref readonly` feature. `in x` must be used with a location (field or local variable). When passing value `x`, the compiler will implicitly assign a location `y = x` and pass `in y`. `ref readonly` disallows such implicit variable.

`this` is a value, not stored a particular location. The same applies to string literals etc. You can't say `in "abc"` either.

**Reply by aa89227** [reply]
Thank you for your explanation. I believe I now understand why it's not possible to use those features in this context. I appreciate your help in clarifying this matter!
